game-core/src/main/java/games/strategy/engine/data/ModifiableProperty.java

```diff
@@ -0,0 +1,145 @@
+package games.strategy.engine.data;
+
+import java.util.function.Supplier;
+
+/**
+ * A wrapper interface to Bundle setters, getters and resetters of the same field.
+ *
+ * @param <T> The type of the field to set, get and reset.
+ */
+public interface ModifiableProperty<T> {
+
+  void setValue(String string) throws Exception;
+
+  void setValue(T value) throws Exception;
+
+  @SuppressWarnings("unchecked")
+  default void setObjectValue(final Object o) {
+    try {
+      if (o instanceof String) {
+        setValue((String) o);
+      } else {
+        setValue((T) o);
+      }
+    } catch (final Exception e) {
+      throw new IllegalStateException("Failed to set Attachment property", e);
+    }
+  }
+
+  T getValue();
+
+  void resetValue();
+
+  /**
+   * Convenience method to create an instance of this interface.
+   */
+  static <T> ModifiableProperty<T> of(
+      final ExceptionConsumer<T> setter,
+      final ExceptionConsumer<String> stringSetter,
+      final Supplier<T> getter,
+      final Runnable resetter) {
+    return new ModifiableProperty<T>() {
+
+      @Override
+      public void setValue(String string) throws Exception {
+        stringSetter.accept(string);
+      }
+
+      @Override
+      public void setValue(T value) throws Exception {
+        setter.accept(value);
+      }
+
+      @Override
+      public T getValue() {
+        return getter.get();
+      }
+
+      @Override
+      public void resetValue() {
+        resetter.run();
+      }
+    };
+  }
+
+  /**
+   * Convenience method to create an instance of this interface with no resetter.
+   */
+  static <T> ModifiableProperty<T> of(
+      final ExceptionConsumer<T> setter,
+      final ExceptionConsumer<String> stringSetter,
+      final Supplier<T> getter) {
+    return of(setter, stringSetter, getter, () -> throwIllegalStateException("No Resetter"));
+  }
+
+  /**
+   * Convenience method to create a generic String instance of this interface.
+   */
+  static ModifiableProperty<String> of(
+      final ExceptionConsumer<String> setter,
+      final Supplier<String> getter,
+      final Runnable resetter) {
+    return of(setter, setter, getter, resetter);
+  }
+
+  /**
+   * Convenience method to create an instance of this interface that only gets.
+   */
+  static <T> ModifiableProperty<T> of(final Supplier<T> getter) {
+    return of(
+        t -> throwIllegalStateException("No Setter"),
```

Comment:
I may have missed your answer to my question in the previous PR, but what's your opinion on throwing `UnsupportedOperationException` here instead?  That seems more appropriate than `IllegalStateException`.

Reply:
Agreed